subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/BuildOperationReportingCompiler.java

```diff
@@ -0,0 +1,95 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile;
+
+import org.gradle.api.internal.TaskInternal;
+import org.gradle.api.internal.tasks.compile.incremental.processing.AnnotationProcessingResult;
+import org.gradle.api.tasks.WorkResult;
+import org.gradle.internal.operations.BuildOperationContext;
+import org.gradle.internal.operations.BuildOperationDescriptor;
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.internal.operations.CallableBuildOperation;
+import org.gradle.language.base.internal.compile.CompileSpec;
+import org.gradle.language.base.internal.compile.Compiler;
+
+import java.util.Collections;
+import java.util.Map;
+
+public class BuildOperationReportingCompiler<T extends CompileSpec> implements Compiler<T> {
+
+    private final Compiler<T> delegate;
+    private final TaskInternal task;
+    private final BuildOperationExecutor buildOperationExecutor;
+
+    public BuildOperationReportingCompiler(TaskInternal task, Compiler<T> delegate, BuildOperationExecutor buildOperationExecutor) {
+        this.delegate = delegate;
+        this.task = task;
+        this.buildOperationExecutor = buildOperationExecutor;
+    }
+
+    @Override
+    public WorkResult execute(final T spec) {
+        return buildOperationExecutor.call(new CallableBuildOperation<WorkResult>() {
+            @Override
+            public BuildOperationDescriptor.Builder description() {
+                return BuildOperationDescriptor.displayName("Invoke compiler for " + task.getIdentityPath())
+                    .details(new Details(task));
+            }
+
+            @Override
+            public WorkResult call(BuildOperationContext context) {
+                WorkResult result = delegate.execute(spec);
+                context.setResult(toBuildOperationResult(result));
+                return result;
+            }
+
+            private Result toBuildOperationResult(WorkResult result) {
+                if (result instanceof CompilationWithAnnotationProcessingResult) {
+                    AnnotationProcessingResult annotationProcessingResult = ((CompilationWithAnnotationProcessingResult) result).getAnnotationProcessingResult();
+                    return new Result(annotationProcessingResult.getExecutionTimeByProcessor());
+                }
+                return new Result(Collections.<String, Long>emptyMap());
```

Comment:
This is not quite correct - There may have been processing, we just don't know (i.e. a non-instrumented compiler was used). We should express this in the result object, so the consumer knows whether to display "no processing" or "don't know if there was processing".

Reply:
I changed that and added tests for forked compilation in f1ecfc146c4.